typing/typecore.mli

```diff
@@ -49,7 +66,7 @@ val check_partial:
         Location.t -> Typedtree.case list -> Typedtree.partial
 val type_expect:
         ?in_function:(Location.t * type_expr) ->
-        Env.t -> Parsetree.expression -> type_expr -> Typedtree.expression
+        Env.t -> Parsetree.expression -> type_expected -> Typedtree.expression
```

Comment:
Wouldn't it be lighter to pass the explanation as an (optional) extra argument, instead of wrapping it in a record with the expected type?

Reply:
Well, that's what I was doing at first, but then @gasche suggested that I attach the explanation to the expected type...

Reply:
More precisely, commit 12d26f3 implements @gasche suggestion, so the commits before that correspond to what you mention.

Reply:
Ah ok, sorry, I did not see @gasche 's comment.  I've a preference for the old version, but I don't mind if the new one is preferred.  If we go for the record solution: (i) would it make sense to include in_function in it?  (ii) perhaps use a function to build the record instead of using literal; it will simplify adding more optional fields later if needed.

Reply:
My intuition is that we will more towards more and more structured representations of the bidirectional information propagated during type-checking, and that having it packed in one value is the right way for nice code evolution.

Reply:
I just pushed a commit that provides a function to build the record as @alainfrisch suggested.